### PR TITLE
Fix for scan output costing

### PR DESCRIPTION
**Pull request by zellerh**
When we cost a scan, the cost of actually producing the output data is added to
the parent node in the query tree. See method CCostModelGPDB::CostChildren().
There is a special case in this method for a filter on top of a scan.

In this case, the cost of producing the output is based on the number of rows produced
by the filter, not the full scan. This code has some issues when we have a
filter on top of a partition selector on top of a scan. In this case, the cost
of producing the outputs is based on the full table scan and that cost is added
to the cost of the partition selector.

Our fix:

In method CCostModelGPDB::CostChildren(), recognize the pattern
filter(partition_selector(scan)) and cost it appropriately.

### CODE REVIEW

File: libgpdbcost/src/CCostModelGPDB.cpp [hunk 1]
```diff
@@ -334,18 +335,49 @@ CCostModelGPDB::CostChildren
...
 		{
 			// by default, compute scan output cost based on full Scan
 			DOUBLE dScanRows = pci->PdRows()[ul];
+			COperator *scanOp = popChild;
 
 			if (fFilterParent)
 			{
-				// if parent is filter, compute scan output cost based on rows produced by Filter operator
-				dScanRows = pci->Rows();
+				CPhysicalPartitionSelector *ps = dynamic_cast<CPhysicalPartitionSelector *>(popChild);
```
Comment (bhuvnesh2703) [inline libgpdbcost/src/CCostModelGPDB.cpp]: why don't we just use if ` COperator::EopPhysicalPartitionSelector == popChild->Eopid()`

File: libgpdbcost/src/CCostModelGPDB.cpp [hunk 2]
```diff
@@ -334,18 +335,49 @@ CCostModelGPDB::CostChildren
...
+
+				if (ps)
+				{
+					CCostContext *grandchildContext = NULL;
+
+					scanOp = exprhdl.PopGrandchild(ul,0,&grandchildContext);
+					CPhysicalDynamicScan *scan = dynamic_cast<CPhysicalDynamicScan*>(scanOp);
```
Comment (bhuvnesh2703) [inline libgpdbcost/src/CCostModelGPDB.cpp]: should we use CPhysicalDynamicScan::PopConvert(scanOp)?

File: data/dxl/minidump/PartSelectorOnJoinSide.mdp [hunk 3]
```diff
@@ -0,0 +1,1795 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
+  <dxl:Comment><![CDATA[
+	Objective: The partition selector should be on inner side of join.
```
Comment (bhuvnesh2703) [inline data/dxl/minidump/PartSelectorOnJoinSide.mdp]: what do we expect if the partition selector is on inner side here?

File: data/dxl/minidump/PartSelectorOnJoinSide.mdp [hunk 4]
```diff
@@ -0,0 +1,1795 @@
...
+	create table t1w(a int, b int, c int, d varchar(10000)) partition by range (b) (start (1) end (10) every (1), default partition def);
+	create table t2w(a int, b int, c int, d varchar(10000));
+
+	insert into t1w select x, x%10, x%100, repeat('x',10000) from (select * from generate_series(1,100000)) T(x);
+	insert into t2w select * from t1w;
+
+	analyze t1w;
+	analyze t2w;
+
+	explain select * from t1w join t2w on t1w.a=t2w.a and t1w.b=t2w.b where t2w.c < 74;
```
Comment (bhuvnesh2703) [inline data/dxl/minidump/PartSelectorOnJoinSide.mdp]: can you paste the plan too?

File: data/dxl/minidump/PartSelectorOnJoinSide2.mdp [hunk 5]
```diff
@@ -0,0 +1,1811 @@
...
+  create table t1n(a int, b int, c int, d varchar(10000)) partition by range (b) (start (1) end (10) every (1), default partition def);
+  create table t2n(a int, b int, c int, d varchar(10000)) partition by range (b) (start (1) end (10) every (1), default partition def);
+
+  insert into t1n select x, x%10, x%100, repeat('x',1000) from (select * from generate_series(1,2000000)) T(x);
+  insert into t2n select * from t1n;
+
+  analyze t1n;
+  analyze t2n;
+
+  explain select t1n.a, t2n.a from t1n join t2n on t1n.a=t2n.a and t1n.b=t2n.b where t2n.c < 74;
```
Comment (bhuvnesh2703) [inline data/dxl/minidump/PartSelectorOnJoinSide2.mdp]: please paste the plan too.

**Comment by zellerh** [comment]
About perf pipeline: We ran it and found no plan change caused by this PR - we did find an issue related to another commit, however. That change has since been reverted.

File: libgpdbcost/src/CCostModelGPDB.cpp (hunk 2 above)
Comment (zellerh) [reply libgpdbcost/src/CCostModelGPDB.cpp]: CPhysicalDynamicScan::PopConvert(scanOp) asserts that the argument is actually a scan, but in this context, we may see operators other than a scan, so we won't be able to use this method.

File: libgpopt/src/operators/CExpressionHandle.cpp [hunk 6]
```diff
@@ -1497,6 +1497,57 @@ CExpressionHandle::Pop
...
+ ULONG grandchild_index,
+ CCostContext **grandchildContext
+ )
+const
+{
+	GPOS_ASSERT(child_index < Arity());
+
+	if (NULL != m_pexpr)
+	{
+		CExpression *childExpr = (*m_pexpr)[child_index];
```
[the remark this answers is not in the document]
Comment (zellerh) [reply libgpopt/src/operators/CExpressionHandle.cpp]: Added the first assert. For the second one, we added an initialization of *grandchildContext if the caller passes a grandchildContext. We expect the caller to handle the case where *grandchildContext == NULL on return.

File: libgpopt/src/operators/CExpressionHandle.cpp [hunk 7]
```diff
@@ -1497,6 +1497,57 @@ CExpressionHandle::Pop
 	return NULL;
 }
 
+COperator *
+CExpressionHandle::PopGrandchild
+(
```
[the remark this answers is not in the document]
Comment (zellerh) [reply libgpopt/src/operators/CExpressionHandle.cpp]: Done

File: libgpopt/src/operators/CExpressionHandle.cpp [hunk 8]
```diff
@@ -1497,6 +1497,57 @@ CExpressionHandle::Pop
...
+ ULONG child_index,
+ ULONG grandchild_index,
+ CCostContext **grandchildContext
+ )
+const
+{
+	GPOS_ASSERT(child_index < Arity());
+
+	if (NULL != m_pexpr)
```
[the remark this answers is not in the document]
Comment (zellerh) [reply libgpopt/src/operators/CExpressionHandle.cpp]: Right now, since we only call CExpressionHandle::PopGrandchild from the costing code, this probably won't ever be true, but we wanted the new method to work in other situations as well.

File: libgpdbcost/src/CCostModelGPDB.cpp [hunk 9]
```diff
@@ -334,18 +335,49 @@ CCostModelGPDB::CostChildren
...
 
 			if (fFilterParent)
 			{
-				// if parent is filter, compute scan output cost based on rows produced by Filter operator
-				dScanRows = pci->Rows();
+				CPhysicalPartitionSelector *ps = dynamic_cast<CPhysicalPartitionSelector *>(popChild);
+
+				if (ps)
+				{
+					CCostContext *grandchildContext = NULL;
```
[the remark this answers is not in the document]
Comment (zellerh) [reply libgpdbcost/src/CCostModelGPDB.cpp]: We talked about this, hope the current names are ok?

File: data/dxl/minidump/PartSelectorOnJoinSide2.mdp (hunk 5 above)
Comment (zellerh) [reply data/dxl/minidump/PartSelectorOnJoinSide2.mdp]: Done

File: data/dxl/minidump/PartSelectorOnJoinSide.mdp (hunk 4 above)
Comment (zellerh) [reply data/dxl/minidump/PartSelectorOnJoinSide.mdp]: Done

**Comment by hardikar** [comment]
I feel it's a little weird that a Filter node affects the number of tuples of its grandchild. IMO, a Filter should affect only its *direct* child.

Is there a reason ORCA wants to produce a plan that has a PS in between a Filter and Scan? That is why can't ORCA produce a plan with a PS on top of the Filter?
Also, what about when there are multiple partition selectors (for different scans)? For example:
```
Filter
 -> PS scan 1
      -> PS scan 2
           -> Scan
```
I think such a situation is possible - It'll take me a bit to return with a working example.

I feel like it's probably better to explore if we can *push* the Filter down to the Scan. Since as it stands this PR also loses the cost of the PS itself (according to your comments).

**Comment by hardikar** [comment]
Ok I gave it a shot - but I was unsuccessful (skip to end for summary)
```
CREATE TABLE rank1 (id int, gender char(1)) DISTRIBUTED BY (id) PARTITION BY LIST (gender) ( PARTITION girls VALUES ('F'), PARTITION boys VALUES ('M'), DEFAULT PARTITION other );
CREATE TABLE rank2 (id int, gender char(1)) DISTRIBUTED BY (id) PARTITION BY LIST (gender) ( PARTITION girls VALUES ('F'), PARTITION boys VALUES ('M'), DEFAULT PARTITION other );
CREATE TABLE other (id int, a char(1));

-- insert into others select i, 'M' from generate_series(1, 100000) i;
-- insert into rank2 select 6, 'M' from generate_series(1, 100000) i;
set optimizer_enumerate_plans = 1;
set optimizer_plan_id = 1;

explain select * from rank1, other, rank2 where rank1.gender = a and rank2.gender = a and rank2.id = 6;
```

The plan has the following:
```
|--CPhysicalPartitionSelector, Scan Id: 1, Part Table: (32815,1.0)  origin: [Grp:6, GrpExpr:3]
|  +--CPhysicalMotionBroadcast    origin: [Grp:6, GrpExpr:4]
|     +--CPhysicalFilter   origin: [Grp:6, GrpExpr:1]
|        |--CPhysicalPartitionSelector, Scan Id: 2, Part Table: (32830,1.0)   origin: [Grp:2, GrpExpr:2]
|        |  +--CPhysicalDynamicTableScan "rank2" ("rank2"), 
```
Note that the first PS & Filter & Motion are in the same group. That is, if it weren’t for the Motion there, the Filter & PS could be reversed (in theory). In practice though, I suppose that won’t happen, since I wasn’t able to get it to. If I remove the Motion node (by adding an equality predicate on ids), the number of alternatives generated is too high for me to go through.

TLDR; I tried to reproduce the case I mentioned in my previous comment, but I wasn’t able to after spending a few hours. In theory, it looks to be possible, but there might be something preventing it. Nevertheless, I feel that Filter should be the direct parent of the Scan, since that is how the plan is translated anyway. Thoughts?

**Comment by zellerh** [comment]
> Nevertheless, I feel that Filter should be the direct parent of the Scan, since that is how the plan is translated anyway. Thoughts?

That was our first thought. We wanted to fix how partition selector was added. But it is added as an enforcer, and we only allow that directly on top of a scan, so it goes below the filter. Our thought was the same as yours: A good fix might be to push the filter into the scan and then we would get ps(scan + filter). That would allow us to cost each operator independently and not to burden the parent operator with some cost of the scan, but we guess that fix won't happen soon.
